torch_xla/csrc/ops/adaptive_avg_pool2d.cpp

```diff
@@ -0,0 +1,50 @@
+#include "ops/adaptive_avg_pool2d.h"
+#include "lowering_context.h"
+#include "ops/infer_output_shape.h"
+#include "pooling.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "tensorflow/compiler/xla/xla_client/util.h"
+
+namespace torch_xla {
+namespace ir {
+namespace ops {
+namespace {
+
+xla::Shape NodeOutputShape(
+    const Value& input,
+    tensorflow::gtl::ArraySlice<const xla::int64> output_size) {
+  auto lower_for_shape_fn =
+      [output_size](tensorflow::gtl::ArraySlice<const xla::XlaOp> operands)
+      -> xla::XlaOp {
+    XLA_CHECK_EQ(operands.size(), 1);
+    return BuildAdaptiveAvgPool2d(operands[0], output_size);
+  };
+  return InferOutputShape({input.shape()}, lower_for_shape_fn);
+}
+
+}  // namespace
+
+AdaptiveAvgPool2d::AdaptiveAvgPool2d(
+    const Value& input,
+    tensorflow::gtl::ArraySlice<const xla::int64> output_size)
+    : Node(ir::OpKind(at::aten::adaptive_avg_pool2d), {input},
+           NodeOutputShape(input, output_size),
+           /*num_outputs=*/1, xla::util::MHash(output_size)),
+      output_size_(output_size.begin(), output_size.end()) {}
+
+XlaOpVector AdaptiveAvgPool2d::Lower(LoweringContext* loctx) const {
+  xla::XlaOp input = loctx->GetOutputOp(operand(0));
+  xla::XlaOp output = BuildAdaptiveAvgPool2d(input, output_size_);
+  return ReturnOp(output, loctx);
+}
+
+std::string AdaptiveAvgPool2d::ToString() const {
+  std::stringstream ss;
+  ss << Node::ToString() << ", output_size=["
+     << absl::StrJoin(output_size_, ", ") << "]";
```

Comment:
I am pretty sure there is a streaming operator for this, but if not, this is getting repeated enough we should have one.

Reply:
Streaming for what, `output_size_`? It's an `std::vector`, not sure what can we do about it.

Reply:
We can write a streamer for it.
But before that, can you try to just stream the vector?

Reply:
Nope, it doesn't work, fails to compile with `no match for ‘operator<<’` error. If we add it, it'd have to be in the `std` namespace I think, which we shouldn't do.

Reply:
Leave it, I will do it later.

http://cpp.sh/9pnor


template <typename T>
std::ostream& operator<< (std::ostream& stream, const std::vector<T>& values) {
    for (const auto& v : values) {
        stream << v << ",";   
    }
    return stream;   
}

Reply:
I'm not sure I'll agree with doing it. Dropping a streaming operator overload in either `std` or global namespace isn't a good solution.

Reply:
And repeating all over the same code is indeed a good solution :)

Reply:
It is not, but polluting either the global or the std namespace with an overload of the streaming operator for an STL type used everywhere is worse.

Reply:
I am not sure what gave you the impression it has to be global or std ns:

http://cpp.sh/4pv5o

Reply:
Oh right, ADL. Yeah, we can do that.